File: languages/dotnet/tests/Revaly.Sdk.Tests/QuickstartTests.cs
using Revaly.Sdk.Errors;
using Revaly.Sdk.Reconcile;
using Revaly.Sdk.Testing;
using Revaly.Sdk.Tests.TestSupport;

namespace Revaly.Sdk.Tests;

/// <summary>
/// Mirrors the quickstart (languages/dotnet/README.md) against the mock transport —
/// the DX-contract §d requirement that our own quickstart tests use the mock. If this
/// test needs to change shape, the README changes with it.
/// </summary>
public class QuickstartTests
{
    private static async Task<string> RunQuickstartFlow(RapClient client, string merchantTransactionId)
    {
        try
        {
            var response = await client.Payments.ChargePaymentAsync(
                TestClient.ChargeRequest(merchantTransactionId));
            response.TryOk(out var transaction);
            return $"approved:{transaction!.TransactionId}";
        }
        catch (PermanentRejectionException ex)
        {
            // Fix or decline. Never fail over — the same request fails anywhere.
            return $"rejected:{ex.StatusCode}";
        }
        catch (TransientFailureException)
        {
            // Definitively not processed — route to your own gateway immediately.
            return "failover";
        }
        catch (OutcomeUnknownException)
        {
            // May have been processed — reconcile before acting.
            var verdict = await client.ReconcileAsync(
                merchantTransactionId,
                new ReconcilePolicy(
                    maxAttempts: 5,
                    overallBudget: TimeSpan.FromSeconds(30),
                    initialDelay: TimeSpan.FromMilliseconds(1)));

            return verdict switch
            {
                FoundVerdict { Outcome: RapTransactionOutcome.Approved } => "reconciled-approved",
                FoundVerdict { IsTerminal: true } => "reconciled-terminal",
                FoundVerdict => "reconciled-keep-polling",
                NotFoundYetVerdict => "hold-and-escalate",
                // The verdict set is open for extension (SafeToFailover arrives with
                // P-2): unknown verdicts are treated conservatively.
                _ => "hold-and-escalate",
            };
        }
    }

    [Fact]
    public async Task Success_path()
    {
        var mock = new RapMockTransport();
        mock.Charge().ReturnsApproved();
        using var client = TestClient.Create(mock);

        Assert.Equal($"approved:{SyntheticData.TransactionId}", await RunQuickstartFlow(client, SyntheticData.MerchantTransactionId));
    }

    [Fact]
    public async Task Rejection_path()
    {
        var mock = new RapMockTransport();
        mock.Charge().ReturnsRejected(422);
        using var client = TestClient.Create(mock);

        Assert.Equal("rejected:422", await RunQuickstartFlow(client, SyntheticData.MerchantTransactionId));
    }

    [Fact]
    public async Task Fast_failover_path()
    {
        var mock = new RapMockTransport();
        mock.Charge().ReturnsNotProcessed503();
        using var client = TestClient.Create(mock);

        Assert.Equal("failover", await RunQuickstartFlow(client, SyntheticData.MerchantTransactionId));
    }

    [Fact]
    public async Task Outcome_unknown_then_reconciled_approved_path()
    {
        var mock = new RapMockTransport();
        mock.Charge().ReturnsTimeoutAfterSend();
        mock.Reconcile(SyntheticData.MerchantTransactionId).ReturnsNotFoundYet().ReturnsApproved();
        using var client = TestClient.Create(mock);

        Assert.Equal("reconciled-approved", await RunQuickstartFlow(client, SyntheticData.MerchantTransactionId));
    }

    [Fact]
    public async Task Outcome_unknown_then_sustained_absence_path()
    {
        var mock = new RapMockTransport();
        mock.Charge().Returns500();
        mock.Reconcile(SyntheticData.MerchantTransactionId).ReturnsNotFoundYet();
        using var client = TestClient.Create(mock);

        Assert.Equal("hold-and-escalate", await RunQuickstartFlow(client, SyntheticData.MerchantTransactionId));
    }
}

File: languages/java/tests/src/test/java/co/revaly/sdk/QuickstartTests.java
package co.revaly.sdk;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.junit.jupiter.api.Assertions.fail;

import co.revaly.sdk.core.model.TransactionResponse;
import co.revaly.sdk.errors.OutcomeUnknownException;
import co.revaly.sdk.errors.PermanentRejectionException;
import co.revaly.sdk.errors.TransientFailureException;
import co.revaly.sdk.reconcile.RapReconcileVerdict;
import co.revaly.sdk.reconcile.RapTransactionOutcome;
import co.revaly.sdk.reconcile.ReconcilePolicy;
import co.revaly.sdk.testing.RapMockTransport;
import java.time.Duration;
import org.junit.jupiter.api.Test;

/**
 * The README quickstart, executed: install → init → charge → handle all three error classes →
 * reconcile with every verdict branch including the default (DX contract §b — these tests keep the
 * copy-paste example honest).
 */
class QuickstartTests {

    @Test
    void quickstartHappyPathChargesSuccessfully() throws Exception {
        RapMockTransport mock = new RapMockTransport();
        mock.charge().returnsApproved();

        RapClient client =
                RapClient.builder()
                        .apiKey(TestClient.API_KEY)
                        .baseUrl("https://sandbox.synthetic.test")
                        .transport(mock)
                        .build();

        TransactionResponse response = client.charge(TestClient.paymentRequest());
        assertEquals(1, response.getTransactionStatus());
    }

    @Test
    void quickstartBranchesOnAllThreeErrorClasses() {
        boolean permanentHandled = false;
        boolean transientHandled = false;
        boolean unknownHandled = false;

        // PermanentRejection: fix or decline — never fail over.
        RapMockTransport rejected = new RapMockTransport();
        rejected.charge().returnsPermanentRejection(422);
        try {
            TestClient.client(rejected).charge(TestClient.paymentRequest());
            fail("expected a typed failure");
        } catch (PermanentRejectionException e) {
            permanentHandled = true;
        } catch (Exception e) {
            fail("wrong class: " + e);
        }

        // TransientFailure: definitively not processed — fail over immediately.
        RapMockTransport breakerOpen = new RapMockTransport();
        breakerOpen.charge().returnsNotProcessed503();
        try {
            TestClient.client(breakerOpen).charge(TestClient.paymentRequest());
            fail("expected a typed failure");
        } catch (TransientFailureException e) {
            transientHandled = true;
        } catch (Exception e) {
            fail("wrong class: " + e);
        }

        // OutcomeUnknown: may have been processed — reconcile before acting.
        RapMockTransport ambiguous = new RapMockTransport();
        ambiguous.charge().throwsTimeoutAfterSend();
        try {
            TestClient.client(ambiguous).charge(TestClient.paymentRequest());
            fail("expected a typed failure");
        } catch (OutcomeUnknownException e) {
            unknownHandled = true;
        } catch (Exception e) {
            fail("wrong class: " + e);
        }

        assertTrue(permanentHandled && transientHandled && unknownHandled);
    }

    @Test
    void quickstartReconcileWorkedExampleAvoidsTheDoubleCharge() throws Exception {
        RapMockTransport mock = new RapMockTransport();
        mock.charge().throwsTimeoutAfterSend();
        mock.reconcile(TestClient.MTX).notFoundYet(1).thenFoundApproved();
        RapClient client = TestClient.client(mock);

        boolean failedOver = false;
        try {
            client.charge(TestClient.paymentRequest());
            fail("expected a typed failure");
        } catch (OutcomeUnknownException e) {
            RapReconcileVerdict verdict =
                    client.reconcile(
                            TestClient.MTX,
                            ReconcilePolicy.builder()
                                    .maxAttempts(5)
                                    .overallBudget(Duration.ofSeconds(10))
                                    .initialDelay(Duration.ofMillis(1))
                                    .build());

            if (verdict instanceof RapReconcileVerdict.Found) {
                RapReconcileVerdict.Found found = (RapReconcileVerdict.Found) verdict;
                if (found.getOutcome() == RapTransactionOutcome.APPROVED) {
                    // The payment already succeeded at RAP-core: failing over here
                    // would charge the cardholder twice.
                    failedOver = false;
                } else {
                    fail("scripted approved");
                }
            } else if (verdict instanceof RapReconcileVerdict.NotFoundYet) {
                fail("scripted found on the second poll");
            } else {
                // Default branch — REQUIRED: verdicts are open for extension.
                fail("future verdict type reached the example");
            }
        }

        assertFalse(failedOver);
    }
}
